### ISSUE TITLE
[Az.Functions] New-AzFunctionAppPlan needs to support creating consumption plan

### ISSUE THREAD

**Issue by Francisco-Gamino** [Azure/azure-powershell#11566]
## Description
New-AzFunctionAppPlan needs to support creating consumption plans. We need this to let Functions customers switch between Premium and Consumption and hosting options.

**Comment by Francisco-Gamino** [comment]
cc: @eamonoreilly

**Comment by jschwanz** [comment]
+1 this is a blocking issue for us.

**Comment by Francisco-Gamino** [comment]
Hello @jschwanz -- could you please describe your scenario? If you are trying to create a consumption app, you can run the `New-AzFunctionApp` cmdlet specifying the `-Location` parameter, which is the geographical location where the function app will be hosted in a consumption plan. With this option, the backend creates a consumption plan underneath for your function app.
From here, you have the option to upgrade from consumption to Premium. If you'd like to do that, you can create a new Premium plan using the `New-AzFunctionAppPlan` cmdlet. Once the plan is created, you can use the `Update-AzFunctionApp` cmldet using the `-PlanName <the created Premium plan name>` to update the function app hosting plan. 
In summary, you can create a consumption function app and then move it a Premium plan. However, going the other way is not supported. Lastly, make sure you run `Get-AzFunctionAppAvailableLocation` for both `-PlanType Consumption` and  `Premium` to find a location where both Premium and Consumption are supported (so that later you can upgrade your function app hosting plan).

**Comment by jschwanz** [comment]
Our users need to transition from Consumption (Dynamic - Y1) to Premium (EP#) and back when the situation calls for it. We also have existing functions on Dedicated plans that we'll want to convert over to Premium. Historically we've been able to make these sku changes by simply calling Set-AzAppServicePlan, but the rigidity of the *-AzFunctionAppPlan cmdlets around Premium-only makes things more difficult than it needs to be.

**Comment by Francisco-Gamino** [comment]
Adding @anthonychu @eamonoreilly @anirudhgarg 

Customer would like to change plan sku from dynamic (Dynamic - Y1) to Premium (EP#) and vise-versa. Currently, New/Update-AzFunctionAppPlan only support Premium plans.

**Comment by Ayanmullick** [comment]
@Francisco-Gamino, One should be able to adhere to one's naming convention while creating a Consumption tier Function app plan too. [This Example](https://docs.microsoft.com/en-us/powershell/module/az.functions/new-azfunctionapp?view=azps-5.7.0#example-1--create-a-consumption-powershell-function-app-in-central-us-) creates a Function app plan with the name `CentralUSPlan`. And [`Update-AzFunctionAppPlan`](https://docs.microsoft.com/en-us/powershell/module/az.functions/update-azfunctionappplan?view=azps-5.7.0) has no way of renaming it either.

Trying a workaround with an "AppServicePlan" fails too

```ps
$Name= '<>'
$Location='NorthCentralUS'
New-AzResourceGroup -Location $Location -Name $Name'RG' 
New-AzAppServicePlan -Name $Name -Location $Location -ResourceGroupName $Name'RG'  -Tier Y1  
New-AzFunctionApp -OSType Windows -Runtime PowerShell -ResourceGroupName $Name'RG' -Name $Name -StorageAccountName <> -PlanName $Name -SubscriptionId <> -DisableApplicationInsights
```

`New-AzAppServicePlan` runs successfully and creates a consumption ASP. However, the last line fails to point to that ASP while creating a FunctionApp.
![image](https://user-images.githubusercontent.com/17693249/114321382-2f30a100-9ae0-11eb-956e-dc75959444cf.png)

**Comment by Ayanmullick** [comment]
> @Francisco-Gamino, One should be able to adhere to one's naming convention while creating a Consumption tier Function app plan too. [This Example](https://docs.microsoft.com/en-us/powershell/module/az.functions/new-azfunctionapp?view=azps-5.7.0#example-1--create-a-consumption-powershell-function-app-in-central-us-) creates a Function app plan with the name `CentralUSPlan`. And [`Update-AzFunctionAppPlan`](https://docs.microsoft.com/en-us/powershell/module/az.functions/update-azfunctionappplan?view=azps-5.7.0) has no way of renaming it either.
> 
> Trying a workaround with an "AppServicePlan" fails too
> 
> ```postscript
> $Name= '<>'
> $Location='NorthCentralUS'
> New-AzResourceGroup -Location $Location -Name $Name'RG' 
> New-AzAppServicePlan -Name $Name -Location $Location -ResourceGroupName $Name'RG'  -Tier Y1  
> New-AzFunctionApp -OSType Windows -Runtime PowerShell -ResourceGroupName $Name'RG' -Name $Name -StorageAccountName <> -PlanName $Name -SubscriptionId <> -DisableApplicationInsights
> ```
> 
> `New-AzAppServicePlan` runs successfully and creates a consumption ASP. However, the last line fails to point to that ASP while creating a FunctionApp.
> ![image](https://user-images.githubusercontent.com/17693249/114321382-2f30a100-9ae0-11eb-956e-dc75959444cf.png)
> 
> Naming one's Consumption Function App plan is doable using [ARM](https://stackoverflow.com/questions/54443700/creating-azure-function-app-against-existing-consumption-plan), [Terraform](https://registry.terraform.io/providers/hashicorp/azurerm/latest/docs/resources/function_app#example-usage-in-a-consumption-plan), AzCLI etc. It should be doable using PowerShell too.


Deploying the Function app without specifying the `-planname` parameter worked.
```ps
$Name= '<>'
$Location='NorthCentralUS'
New-AzResourceGroup -Location $Location -Name $Name'RG' 
New-AzAppServicePlan -Name $Name -Location $Location -ResourceGroupName $Name'RG' -Tier Y1 -Verbose
New-AzFunctionApp -OSType Windows -Runtime PowerShell -ResourceGroupName $Name'RG' -Name $Name -StorageAccountName <> -SubscriptionId <> -DisableApplicationInsights
```

**Comment by Francisco-Gamino** [comment]
Hello @Ayanmullick -- When you create a Consumption function app, the user does not have the option to name the (Consumption) plan. 

To create a function app via the New-AzFunctionApp cmdlet, the user must specify a plan name (for Premium or App Service Plan) or a Location where the Consumption plan will be created. 

Hi @gzuber -- Do you know if the Az CLI support renaming function app service plans? 
 
cc: @stefanushinardi @AnatoliB

**Comment by bryanknox** [comment]
My team wants to be able to control the name of the consumption App Service Plan to help us avoid confusion and costly mistakes.

We also want to be able to move a function app between consumption and premium plans when needed without having to jump through special hoops.
